worker/src/task.rs

```diff
@@ -0,0 +1,473 @@
+                    drop(tx);
+                    return ();
+                }
+                task_output = Some(res.unwrap().await);
+            }
+            if tx.send(task_output.unwrap()).is_err() {
+                warn!("Task {name_cloned} with task_id {task_id} output could not be sent: receiver dropped")
+            }
+            ()
+        });
+        Some(self.get_task_handle::<T>(
+            name,
+            tag,
+            task_id,
+            parent_task_id,
+            cloned_token,
+            fut,
+            detached,
+            rx,
+        ))
+    }
+    /// signals shutdown of this executor and any Clones
+    pub async fn shutdown_and_wait(&mut self) {
+        // hang up the channel which will cause the dedicated thread
+        // to quit
+        let shutdown;
+        {
+            let mut state = self.state.lock();
+            shutdown = state.shutdown;
+            state.shutdown = true;
+        }
+        if !shutdown {
+            self.cancel_all_and_wait().await;
+        }
+    }
+    pub async fn stop_task(&self, task_id: u64) {
+        let task;
+        {
+            let mut state = self.state.lock();
+            task = state.tasks.get_mut(&task_id).cloned();
+        }
+        if task.is_none() {
+            error!("Task with task_id {task_id} doesn't exist!");
+        } else {
+            let shutdown = task.unwrap().shutdown;
+            shutdown.clone().shutdown();
+            shutdown.clone().wait_shutdown_complete().await;
+        }
+    }
+    pub async fn stop_named_task(&self, name: String) {
+        let task_id;
+        {
+            let state = self.state.lock();
+            let task = state.tasks_by_name.get(&name);
+            if task.is_none() {
+                error!("Task with name {name} doesn't exist!");
+            }
+            let task = state.tasks.get(task.unwrap());
+            if task.is_none() {
+                panic!("Task missing in task map {name}");
+            } else {
+                task_id = task.unwrap().task_id;
+            }
+        }
+        self.stop_task(task_id);
+    }
+    pub async fn stop_tagged_tasks(&self, tag: String) {
+        let mut task_ids: Vec<u64> = vec![];
+        let mut futures: Vec<Pin<Box<dyn Future<Output = ()>>>> = vec![];
+        {
+            let state = self.state.lock();
+            for (task_id, task) in state.tasks.iter() {
+                if task.tag.is_some() && *task.tag.as_ref().unwrap() == tag {
+                    task_ids.push(*task_id);
+                }
+            }
+
+        }
+        for task_id in task_ids {
+            futures.push(Box::pin(self.stop_task(task_id)));
+        }
+        futures::future::join_all(futures).await;
+    }
+}
+
+impl Drop for TaskTracker {
+    fn drop(&mut self) {
+        let handle = tokio::runtime::Handle::current();
+        handle.enter();
+        futures::executor::block_on(self.cancel_all_and_wait())
```

Comment:
This can cause deadlocks. In particular it is not safe to block on async inside of a drop function. There's a reason why tokio will panic if you try.

Reply:
What will cause deadlock other than when the tokio runtime is running with a single thread which is not how we are running things in prod? The only other thing that i can think of is when someone creates a bunch of task trackers  (TaskTracker) instances (it should ideally be a singleton) which is not how this is supposed to be used as well and will get caught in code reviews. The above code successfully runs in my tests but maybe I am missing the point in "There's a reason why tokio will panic if you try."
In any case, to prevent us from doing the wrong thing even by mistake, I've removed the blocking logic in drop and made it check for shutdown to have been completed beforehand or panic otherwise.